File: src/favorites.ts
// ♥ Favorites (FAVORITES.md, NEXT-VERSION §3) — the front-end half of the love rating.
//
// One in-memory set of loved catalog ids, seeded from the Rust mirror at boot and kept
// in step by three feeds: our own writes (optimistic flip → PUT/DELETE → roll back on
// error), Apple's generated "Favorite Songs" playlist (seeded in Rust whenever its
// tracks are fetched; the Playlists card triggers that once per session), and a
// batched ratings read for ids the mirror has never seen (`reconcile`). Menus and the
// Now Playing ♥ read `isLoved` synchronously and subscribe to `onFavoritesChange`.
//
// Gate: the ♥ is an account write, so it rides the same consent as Add to Library
// (the Library Add toggle) — one switch for "DeetsMusic may write to my Apple account".

import { invoke } from "@tauri-apps/api/core";
import type { Track } from "./library";
import type { MenuItem } from "./context-menu";
import { libraryAddEnabled } from "./library-add";
import { toast } from "./toast";
import { catalogRelated, type Playlist } from "./search";

const loved = new Set<string>();
const known = new Set<string>(); // every id the mirror holds a row for (loved or not)
const asked = new Set<string>(); // ids already sent to reconcile this session
let ready = false;

const listeners = new Set<() => void>();
export function onFavoritesChange(cb: () => void): () => void {
  listeners.add(cb);
  return () => listeners.delete(cb);
}
const emit = () => listeners.forEach((cb) => cb());

/** Load the mirror once at boot (zero Apple calls). Safe to call again after a seed. */
export async function initFavorites(): Promise<void> {
  try {
    const [l, k] = await Promise.all([
      invoke<string[]>("favorites_cached"),
      invoke<string[]>("favorites_known"),
    ]);
    loved.clear();
    l.forEach((id) => loved.add(id));
    k.forEach((id) => known.add(id));
    ready = true;
    emit();
  } catch (e) {
    console.warn("[favorites] load", e);
  }
}

/** Is the ♥ offered for this track at all? (consent on, has a catalog id) */
export const favoriteOffered = (t: Track | undefined): t is Track =>
  !!t?.catalogId && libraryAddEnabled();

export const isLoved = (t: Track | undefined): boolean => !!t?.catalogId && loved.has(t.catalogId);

/** ♥ or un-♥. Optimistic: the set flips now, Apple is told, a failure flips it back. */
export async function setLoved(t: Track, on: boolean): Promise<void> {
  const id = t.catalogId;
  if (!id) throw new Error("track has no catalog id");
  const was = loved.has(id);
  if (on) loved.add(id);
  else loved.delete(id);
  known.add(id);
  emit();
  try {
    await invoke("favorite_set", { track: t, loved: on });
  } catch (e) {
    if (was) loved.add(id);
    else loved.delete(id);
    emit();
    // The ♥ just flipped back; say why (TOASTS.md). Every ♥ path (menus, Now Playing) lands here.
    toast({ kind: "warn", text: `Couldn't update Favorites for “${t.title}”.` });
    throw e;
  }
}

export const toggleLoved = (t: Track): Promise<void> => setLoved(t, !isLoved(t));

/**
 * Ask Apple about the tracks on screen the mirror has never seen — a ♥ set on the
 * phone before Favorite Songs caught up. Batched (100 ids per call in Rust), each id
 * asked at most once per install, so a card render costs at most one small read and
 * usually none. Fire-and-forget; the change bus repaints when answers land.
 */
export function reconcile(tracks: Track[]): void {
  if (!ready || !libraryAddEnabled()) return;
  const ids: string[] = [];
  for (const t of tracks) {
    const id = t.catalogId;
    if (!id || known.has(id) || asked.has(id)) continue;
    asked.add(id);
    ids.push(id);
  }
  if (!ids.length) return;
  invoke<{ loved: string[]; unloved: string[] }>("favorites_reconcile", { ids })
    .then((r) => {
      r.loved.forEach((id) => { loved.add(id); known.add(id); });
      r.unloved.forEach((id) => { loved.delete(id); known.add(id); });
      if (r.loved.length) emit();
    })
    .catch((e) => console.warn("[favorites] reconcile", e));
}

// ── Albums and playlists (CONTEXT-MENUS.md §3b, his calls F1A F2A F3A, 2026-09-24) ──
// The same set, keyed `album:{id}` / `playlist:{id}` (favorites.rs), plus `albumsong:{id}`
// for a Library album, which knows only its songs: the first song with a catalog id stands
// for the album, so its menu reads the state with no song → album lookup. A menu reads the
// set; a hero, when it opens, asks Apple once per install (`reconcileAlbum` /
// `reconcilePlaylist`). An album or playlist never asked reads "Favorite" — pressing it on
// one already loved only loves it again.

/** An album as the ♥ sees it: its catalog id, its first song's catalog id, or both. */
export interface AlbumFav {
  id?: string;
  seed?: string;
  name: string;
}
/** A playlist as the ♥ sees it: one of Apple's (catalog id) or a library one (`p.` id). */
export interface PlaylistFav {
  kind: "playlists" | "library-playlists";
  id: string;
  name: string;
}

/** How Apple rates this playlist: a library one by its `p.` id, one of Apple's by its catalog
 *  id. A playlist made in DeetsMusic has no Apple rating, so no ♥ (his call F3A). */
export function playlistFav(p: Playlist): PlaylistFav | undefined {
  if (p.source === "local") return undefined;
  if (p.libraryId?.startsWith("p.")) return { kind: "library-playlists", id: p.libraryId, name: p.name };
  if (p.catalogId) return { kind: "playlists", id: p.catalogId, name: p.name };
  return undefined;
}

const albumKeys = (a: AlbumFav): string[] =>
  [a.id && `album:${a.id}`, a.seed && `albumsong:${a.seed}`].filter((k): k is string => !!k);
const isAlbumLoved = (a: AlbumFav): boolean => albumKeys(a).some((k) => loved.has(k));
const albumKnown = (a: AlbumFav): boolean => albumKeys(a).some((k) => known.has(k));
const playlistKey = (p: PlaylistFav): string => `playlist:${p.id}`;

/** The album's catalog id: its own, or its first song's album (one session-cached hop). */
async function albumId(a: AlbumFav): Promise<string | undefined> {
  if (a.id) return a.id;
  if (!a.seed) return undefined;
  return (await catalogRelated("songs", a.seed, "albums"))?.id;
}

/** Flip the set now; roll back and say so if Apple refuses (as `setLoved`). */
async function flip(keys: string[], on: boolean, name: string, write: () => Promise<unknown>): Promise<void> {
  const was = keys.map((k) => loved.has(k));
  keys.forEach((k) => (on ? loved.add(k) : loved.delete(k), known.add(k)));
  emit();
  try {
    await write();
  } catch (e) {
    keys.forEach((k, i) => (was[i] ? loved.add(k) : loved.delete(k)));
    emit();
    toast({ kind: "warn", text: `Couldn't update Favorites for “${name}”.` });
    throw e;
  }
}

/** ♥ / un-♥ an album. A Library album finds its catalog id first. */
export async function setAlbumLoved(a: AlbumFav, on: boolean): Promise<void> {
  const id = await albumId(a);
  if (!id) {
    toast({ kind: "warn", text: `Couldn't find “${a.name}” on Apple Music.` });
    return;
  }
  const full = { ...a, id };
  await flip(albumKeys(full), on, a.name, () =>
    invoke("favorite_collection_set", { kind: "albums", id, alias: a.seed ?? null, loved: on }),
  );
}

export async function setPlaylistLoved(p: PlaylistFav, on: boolean): Promise<void> {
  await flip([playlistKey(p)], on, p.name, () =>
    invoke("favorite_collection_set", { kind: p.kind, id: p.id, alias: null, loved: on }),
  );
}

/** The album hero opened: ask Apple once per install whether it is loved. */
export function reconcileAlbum(a: AlbumFav): void {
  if (!ready || !libraryAddEnabled() || albumKnown(a)) return;
  const k = albumKeys(a)[0];
  if (!k || asked.has(k)) return;
  asked.add(k);
  void albumId(a)
    .then((id) =>
      id
        ? invoke<boolean>("favorite_collection_reconcile", { kind: "albums", id, alias: a.seed ?? null }).then((on) => {
            albumKeys({ ...a, id }).forEach((key) => (on ? loved.add(key) : loved.delete(key), known.add(key)));
            if (on) emit();
          })
        : undefined,
    )
    .catch((e) => console.warn("[favorites] album reconcile", e));
}

/** The playlist hero opened: ask Apple once per install whether it is loved. */
export function reconcilePlaylist(p: PlaylistFav): void {
  const k = playlistKey(p);
  if (!ready || !libraryAddEnabled() || known.has(k) || asked.has(k)) return;
  asked.add(k);
  invoke<boolean>("favorite_collection_reconcile", { kind: p.kind, id: p.id, alias: null })
    .then((on) => {
      known.add(k);
      if (on) {
        loved.add(k);
        emit();
      }
    })
    .catch((e) => console.warn("[favorites] playlist reconcile", e));
}

/** Favorite / Unfavorite for an album, or null when not offered (gate off, no id at all). */
export function albumFavoriteItem(a: AlbumFav): MenuItem | null {
  if (!libraryAddEnabled() || (!a.id && !a.seed)) return null;
  const on = isAlbumLoved(a);
  return {
    label: on ? "Unfavorite" : "Favorite",
    run: () => void setAlbumLoved(a, !on).catch((e) => console.error("[favorites] album", e)),
  };
}

/** Favorite / Unfavorite for a playlist, or null (gate off, or no Apple rating: F3A). */
export function playlistFavoriteItem(p: PlaylistFav | undefined): MenuItem | null {
  if (!p || !libraryAddEnabled()) return null;
  const on = loved.has(playlistKey(p));
  return {
    label: on ? "Unfavorite" : "Favorite",
    run: () => void setPlaylistLoved(p, !on).catch((e) => console.error("[favorites] playlist", e)),
  };
}

/** The shared right-click item: Favorite / Unfavorite, or null when not offered. */
export function favoriteItem(t: Track | undefined): MenuItem | null {
  if (!favoriteOffered(t)) return null;
  const on = isLoved(t);
  return {
    label: on ? "Unfavorite" : "Favorite",
    run: () => void setLoved(t, !on).catch((e) => console.error("[favorites] set", e)),
  };
}

File: src-tauri/src/favorites.rs
//! ♥ Favorites (FAVORITES.md, NEXT-VERSION §3): Apple's love rating, love only.
//!
//! Apple has no list-all-favorites endpoint, so a local `favorites` mirror backs the
//! filled state. Three feeds keep it honest: our own writes (write-through), the
//! generated "Favorite Songs" library playlist (the only list-all — seeded whenever
//! that playlist's tracks are fetched, zero extra calls), and a batched
//! `GET …/ratings/songs?ids=` for ids the playlist has not caught up with yet.
//! Keyed catalog-first like everything else; catalog-less uploads cannot be loved.

use crate::lock::LockExt;
use rusqlite::Connection;
use serde::Serialize;
use tauri::State;

use crate::apple::{self, AppleState};
use crate::library::{self, Db};
use crate::model::Track;

fn err<E: std::fmt::Display>(e: E) -> String {
    e.to_string()
}

fn now_s() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

/// Upsert one mirror row.
pub(crate) fn set_local(conn: &Connection, id: &str, loved: bool) -> Result<(), String> {
    conn.execute(
        "INSERT INTO favorites(track_id, loved, synced_at) VALUES(?1, ?2, ?3)
         ON CONFLICT(track_id) DO UPDATE SET loved = excluded.loved, synced_at = excluded.synced_at",
        rusqlite::params![id, loved as i64, now_s()],
    )
    .map_err(err)?;
    Ok(())
}

/// Seed the mirror from Apple's generated **Favorite Songs** playlist: every track in
/// it is loved. Rows not in it are left alone (the playlist can lag a fresh ♥ by a
/// while, and it only lists songs Apple could add to the library). Also parks the
/// tracks as `seen` rows so a favorites view can always resolve metadata.
pub(crate) fn seed_from_playlist(conn: &Connection, tracks: &[Track]) -> Result<usize, String> {
    library::materialize_many(conn, tracks)?;
    let mut n = 0usize;
    for t in tracks {
        let Some(id) = t.catalog_id.clone().filter(|s| !s.is_empty()) else { continue };
        set_local(conn, &id, true)?;
        n += 1;
    }
    Ok(n)
}

/// Is this mirror row Apple's generated Favorite Songs list? There is no type field:
/// the tell is the fixed name on a read-only library playlist with no catalog twin.
pub(crate) fn is_favorite_songs(p: &crate::model::Playlist) -> bool {
    p.name == "Favorite Songs" && !p.can_edit && p.global_id.is_none()
}

/// ♥ / un-♥ one song: `PUT …/ratings/songs/{id}` value 1, or `DELETE`. Write-through:
/// the mirror row flips only after Apple accepts, so the optimistic front-end can
/// roll back on an error. The track is parked as `seen` first (metadata for later).
#[tauri::command]
pub async fn favorite_set(
    track: Track,
    loved: bool,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<(), String> {
    let id = track
        .catalog_id
        .clone()
        .filter(|s| !s.is_empty())
        .ok_or("favorite_set: no catalog id")?;
    let dev = apple::developer_token()?;
    let user = state.user_token.lock_or_recover().clone().ok_or("not connected to Apple Music")?;
    let client = crate::apple::http_client();
    let url = format!("https://api.music.apple.com/v1/me/ratings/songs/{id}");
    let (status, body) = if loved {
        let payload = serde_json::json!({ "type": "rating", "attributes": { "value": 1 } });
        apple::api_send(&client, reqwest::Method::PUT, &dev, &user, &url, Some(&payload)).await?
    } else {
        apple::api_send(&client, reqwest::Method::DELETE, &dev, &user, &url, None).await?
    };
    if !(200..300).contains(&status) {
        return Err(format!("favorite HTTP {status}: {body}"));
    }
    let conn = db.lock();
    library::materialize_many(&conn, std::slice::from_ref(&track))?;
    set_local(&conn, &id, loved)?;
    crate::log::info(&format!("favorites: {} {id}", if loved { "loved" } else { "unloved" }));
    Ok(())
}

// ── Albums and playlists (CONTEXT-MENUS.md §3b, his calls F1A F2A F3A, 2026-09-24) ──
// The same mirror table, keyed `album:{id}` / `playlist:{id}` so no song id can collide.
// A Library album holds no album id (only its songs'), so a row `albumsong:{first song id}`
// mirrors the album's row: the Library menu reads it with no song → album lookup.

/// The ratings path segment for a collection kind, or an error for anything else.
fn collection_kind(kind: &str) -> Result<(&'static str, &'static str), String> {
    match kind {
        "albums" => Ok(("albums", "album")),
        "playlists" => Ok(("playlists", "playlist")),
        "library-playlists" => Ok(("library-playlists", "playlist")),
        _ => Err(format!("favorites: bad collection kind '{kind}'")),
    }
}

fn set_collection_local(conn: &Connection, prefix: &str, id: &str, alias: Option<&str>, loved: bool) -> Result<(), String> {
    set_local(conn, &format!("{prefix}:{id}"), loved)?;
    if let Some(song) = alias.filter(|s| !s.is_empty()) {
        set_local(conn, &format!("albumsong:{song}"), loved)?;
    }
    Ok(())
}

/// ♥ / un-♥ an album or playlist: `PUT …/ratings/{kind}/{id}` value 1, or `DELETE`.
/// Write-through, as `favorite_set`. `alias`: the album's first song, for the Library row.
#[tauri::command]
pub async fn favorite_collection_set(
    kind: String,
    id: String,
    alias: Option<String>,
    loved: bool,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<(), String> {
    let (path, prefix) = collection_kind(&kind)?;
    if id.is_empty() {
        return Err("favorite_collection_set: no id".into());
    }
    let dev = apple::developer_token()?;
    let user = state.user_token.lock_or_recover().clone().ok_or("not connected to Apple Music")?;
    let client = crate::apple::http_client();
    let url = format!("https://api.music.apple.com/v1/me/ratings/{path}/{id}");
    let (status, body) = if loved {
        let payload = serde_json::json!({ "type": "rating", "attributes": { "value": 1 } });
        apple::api_send(&client, reqwest::Method::PUT, &dev, &user, &url, Some(&payload)).await?
    } else {
        apple::api_send(&client, reqwest::Method::DELETE, &dev, &user, &url, None).await?
    };
    if !(200..300).contains(&status) {
        return Err(format!("favorite {path} HTTP {status}: {body}"));
    }
    let conn = db.lock();
    set_collection_local(&conn, prefix, &id, alias.as_deref(), loved)?;
    crate::log::info(&format!("favorites: {} {prefix} {id}", if loved { "loved" } else { "unloved" }));
    Ok(())
}

/// Ask Apple once whether an album or playlist is loved (one `GET …/ratings/{kind}?ids=`).
/// The answer gets a mirror row either way, so it is never asked again on this install.
///
/// A background job (the app asks on its own, APPLE-CALLS.md §3): `apple_calls::BUSY` while
/// Apple's back-off holds, and no row is written, so it asks again next time.
#[tauri::command]
pub async fn favorite_collection_reconcile(
    kind: String,
    id: String,
    alias: Option<String>,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<bool, String> {
    if crate::apple_calls::skip("favorite_collection_reconcile") {
        return Err(crate::apple_calls::BUSY.into());
    }
    crate::apple_calls::background("favorite_collection_reconcile", favorite_collection_reconcile_run(kind, id, alias, state, db)).await
}

async fn favorite_collection_reconcile_run(
    kind: String,
    id: String,
    alias: Option<String>,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<bool, String> {
    let (path, prefix) = collection_kind(&kind)?;
    if id.is_empty() {
        return Err("favorite_collection_reconcile: no id".into());
    }
    let dev = apple::developer_token()?;
    let user = state.user_token.lock_or_recover().clone().ok_or("not connected to Apple Music")?;
    let client = crate::apple::http_client();
    let url = format!("https://api.music.apple.com/v1/me/ratings/{path}?ids={id}");
    let (status, body) = apple::api_get(&client, &dev, &user, &url).await?;
    // 404 = no rating on it; anything else unexpected is an error.
    if status != 200 && status != 404 {
        return Err(format!("ratings {path} HTTP {status}: {body}"));
    }
    let loved = body
        .get("data")
        .and_then(|d| d.as_array())
        .map(|arr| arr.iter().any(|r| r.pointer("/attributes/value").and_then(|v| v.as_i64()) == Some(1)))
        .unwrap_or(false);
    let conn = db.lock();
    set_collection_local(&conn, prefix, &id, alias.as_deref(), loved)?;
    crate::log::info(&format!("favorites: reconciled {prefix} {id}, loved={loved}"));
    Ok(loved)
}

/// Every loved id in the mirror — the front-end's in-memory set at boot.
#[tauri::command]
pub async fn favorites_cached(app: tauri::AppHandle) -> Result<Vec<String>, String> {
    crate::db_thread::run(&app, move |db| {
        let conn = db.lock();
        let mut stmt = conn
            .prepare_cached("SELECT track_id FROM favorites WHERE loved = 1")
            .map_err(err)?;
        let rows = stmt.query_map([], |r| r.get::<_, String>(0)).map_err(err)?;
        rows.collect::<Result<Vec<_>, _>>().map_err(err)
    })
    .await
}

/// Which ids the mirror already knows (loved or not) — so the front-end only asks
/// Apple about the rest.
#[tauri::command]
pub async fn favorites_known(app: tauri::AppHandle) -> Result<Vec<String>, String> {
    crate::db_thread::run(&app, move |db| {
        let conn = db.lock();
        let mut stmt = conn.prepare_cached("SELECT track_id FROM favorites").map_err(err)?;
        let rows = stmt.query_map([], |r| r.get::<_, String>(0)).map_err(err)?;
        rows.collect::<Result<Vec<_>, _>>().map_err(err)
    })
    .await
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Reconciled {
    pub loved: Vec<String>,
    pub unloved: Vec<String>,
}

/// Ask Apple about a batch of catalog ids the mirror has never seen (a ♥ set on the
/// phone). One `GET …/ratings/songs?ids=` per 100 ids; ids Apple omits are not loved.
/// Every asked id gets a mirror row either way, so it is never asked again on this
/// install (a later ♥ elsewhere shows up through the Favorite Songs seed).
///
/// A background job, like `favorite_collection_reconcile` (APPLE-CALLS.md §3).
#[tauri::command]
pub async fn favorites_reconcile(
    ids: Vec<String>,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<Reconciled, String> {
    if crate::apple_calls::skip("favorites_reconcile") {
        return Err(crate::apple_calls::BUSY.into());
    }
    crate::apple_calls::background("favorites_reconcile", favorites_reconcile_run(ids, state, db)).await
}

async fn favorites_reconcile_run(
    ids: Vec<String>,
    state: State<'_, AppleState>,
    db: State<'_, Db>,
) -> Result<Reconciled, String> {
    let ids: Vec<String> = ids.into_iter().filter(|s| !s.is_empty()).collect();
    let mut out = Reconciled { loved: Vec::new(), unloved: Vec::new() };
    if ids.is_empty() {
        return Ok(out);
    }
    let dev = apple::developer_token()?;
    let user = state.user_token.lock_or_recover().clone().ok_or("not connected to Apple Music")?;
    let client = crate::apple::http_client();
    for chunk in ids.chunks(100) {
        let url = format!(
            "https://api.music.apple.com/v1/me/ratings/songs?ids={}",
            chunk.join(",")
        );
        let (status, body) = apple::api_get(&client, &dev, &user, &url).await?;
        // 404 = none of these ids carries a rating; anything else unexpected is an error.
        if status != 200 && status != 404 {
            return Err(format!("ratings HTTP {status}: {body}"));
        }
        let mut loved_here = std::collections::HashSet::new();
        if let Some(arr) = body.get("data").and_then(|d| d.as_array()) {
            for r in arr {
                let v = r.pointer("/attributes/value").and_then(|v| v.as_i64()).unwrap_or(0);
                if let Some(id) = r.get("id").and_then(|i| i.as_str()) {
                    if v == 1 {
                        loved_here.insert(id.to_string());
                    }
                }
            }
        }
        for id in chunk {
            if loved_here.contains(id) {
                out.loved.push(id.clone());
            } else {
                out.unloved.push(id.clone());
            }
        }
    }
    let conn = db.lock();
    for id in &out.loved {
        set_local(&conn, id, true)?;
    }
    for id in &out.unloved {
        set_local(&conn, id, false)?;
    }
    crate::log::info(&format!(
        "favorites: reconciled {} id(s), {} loved",
        out.loved.len() + out.unloved.len(),
        out.loved.len()
    ));
    Ok(out)
}
